From 5bd9d1150033c29244e2396ece8bc92e41fd5cc2 Mon Sep 17 00:00:00 2001
From: Sandor Yu <R01008@freescale.com>
Date: Mon, 18 May 2015 15:28:21 +0800
Subject: [PATCH 1090/1594] MLK-10911: ov4560: Move mclk rate setting before
 clk on

commit 59cac3fab1eea4b640c84e8f78c22170a5cf7207 from
git://git.freescale.com/imx/linux-2.6-imx.git

Move mclk rate setting code before clk on function.
The change is required by imx clk framework,
otherwise clk_set_rate will failed.

Signed-off-by: Sandor Yu <R01008@freescale.com>
---
 drivers/media/platform/mxc/subdev/ov5640.c |   27 +++++++++++++++++++++------
 1 files changed, 21 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/mxc/subdev/ov5640.c b/drivers/media/platform/mxc/subdev/ov5640.c
index 432bcb6..cad8795 100644
--- a/drivers/media/platform/mxc/subdev/ov5640.c
+++ b/drivers/media/platform/mxc/subdev/ov5640.c
@@ -1652,6 +1652,24 @@ static int ov5640_enum_frame_interval(struct v4l2_subdev *sd,
 	return -EINVAL;
 }
 
+static int ov5640_set_clk_rate(void)
+{
+	u32 tgt_xclk;	/* target xclk */
+	int ret;
+
+	/* mclk */
+	tgt_xclk = ov5640_data.mclk;
+	tgt_xclk = min(tgt_xclk, (u32)OV5640_XCLK_MAX);
+	tgt_xclk = max(tgt_xclk, (u32)OV5640_XCLK_MIN);
+	ov5640_data.mclk = tgt_xclk;
+
+	pr_debug("   Setting mclk to %d MHz\n", tgt_xclk / 1000000);
+	ret = clk_set_rate(ov5640_data.sensor_clk, ov5640_data.mclk);
+	if (ret < 0)
+		pr_debug("set rate filed, rate=%d\n", ov5640_data.mclk);
+	return ret;
+}
+
 /*!
  * dev_init - V4L2 sensor init
  * @s: pointer to standard V4L2 device structure
@@ -1668,12 +1686,6 @@ static int init_device(void)
 
 	/* mclk */
 	tgt_xclk = ov5640_data.mclk;
-	tgt_xclk = min(tgt_xclk, (u32)OV5640_XCLK_MAX);
-	tgt_xclk = max(tgt_xclk, (u32)OV5640_XCLK_MIN);
-	ov5640_data.mclk = tgt_xclk;
-
-	pr_debug("   Setting mclk to %d MHz\n", tgt_xclk / 1000000);
-	clk_set_rate(ov5640_data.sensor_clk, ov5640_data.mclk);
 
 	/* Default camera frame rate is set in probe */
 	tgt_fps = ov5640_data.streamcap.timeperframe.denominator /
@@ -1792,6 +1804,9 @@ static int ov5640_probe(struct i2c_client *client,
 		return retval;
 	}
 
+	/* Set mclk rate before clk on */
+	ov5640_set_clk_rate();
+
 	clk_prepare_enable(ov5640_data.sensor_clk);
 
 	ov5640_data.io_init = ov5640_reset;
-- 
1.7.5.4

